src/Lucid/Base.hs

```diff
@@ -106,40 +106,34 @@ instance MFunctor HtmlT where
   hoist f (HtmlT xs) = HtmlT (f xs)
 
 -- | @since 2.9.7
-instance (a ~ (),Monad m) => Semigroup (HtmlT m a) where
-  (<>) = liftM2 mappend
+instance (Semigroup a,Applicative m) => Semigroup (HtmlT m a) where
+  (<>) = liftA2 (<>)
 
 -- | Monoid is right-associative, a la the 'Builder' in it.
-instance (a ~ (),Monad m) => Monoid (HtmlT m a) where
-  mempty  = return mempty
-  mappend = liftM2 mappend
+instance (Monoid a,Applicative m) => Monoid (HtmlT m a) where
```

Comment:
The `()` was to aid type inference of terms. How does this affect it?

Reply:
@chrisdone 

something like `div_ [ class_ "fancyStyleThing" ] mempty` will be of type `Monoid a => HtmlT m a` not `HtmlT m ()`, so it might cause ambiguous type variables

otoh, `foo <> bar <> baz` will force all `foo`, `bar` and `baz` to have same `Monoid a => HtmlT m a` type, so if `a` is nailed otherwise, inference will work.

---

I also don't know how this will work in practice. Have to try on real world project to see.
Personally I don't need either change. (I don't have `HtmlT m a` with non-`Monad`ic `m`, it's either `Reader`, `State` or `Identity`).

Reply:
Yeah, there'd be ambiguous type vars if you pass a term directly to one of the `render` functions with no type annotations in between:

```haskell
foo :: Text
foo = renderText (div_ mempty)
```

You probably know more than me about how likely this is to crop up. Personally I find it more readable to define my `Html` expressions separately, with a type anno, either at the top level or within a `where`.

```haskell
bar :: Html ()
bar = div_ mempty

foo :: Text
foo = renderText bar
```

`Html`'s `IsString` instance tends to peg `a ~ ()` anyway. I think it's probably more common to use `""` than `mempty`:

```haskell
foo :: Text
foo = renderText (div_ "")
```

All that said, I feel much less strongly about the `a ~ ()` constraints than I do about the `Applicative`/`Functor` constraints, so I'd be quite happy to undo that for the purposes of this PR if you prefer.

Reply:
I'd prefer to keep inference working as much as possible if it's all the same, or otherwise make a major version bump to indicate possible breakage downstream.